### ISSUE TITLE
Is Backtesting supported with RL models ?

### ISSUE THREAD

**Issue by escanoro** [freqtrade/freqtrade#8418]
## Describe your environment

  * Operating system: Ubuntu 22.04
  * Python Version: N/A
  * CCXT version:  N/A
  * Freqtrade Version: freqtrade:develop_freqairl docker image (digest: 9e9d4215f087)
  

## Describe the problem:

When backtesting using a RL model (I'm using the provided model ReinforcementLearner), freqtrade 'hangs' and stop working after throwing this warning:

`freqtrade-train_ai_rl_01  | 2023-03-29 15:29:41,206 - freqtrade.freqai.RL.BaseEnvironment - WARNING - add_state_info is not available in backtesting. Deactivating.`

### Steps to reproduce:

  1. Use the freqtrade:develop_freqairl docker image (digest: 9e9d4215f087)
  2. Start a Backtesting on a very trivial strategy
  
### Observed Results:

  *After few seconds of working, a warning is thrown, and freqtrade hangs (It does not exit but hangs), also CPU usage goes to zero.
  * I was expecting that after the warning, freqtrade continues

The documentation says : 

> A second important component of the FreqAI RL implementation is the use of state information. State information is fed into the network at each step, including current profit, current position, and current trade duration. These are used to train the agent in the training environment, and to reinforce the agent in dry/live (this functionality is not available in backtesting). FreqAI + Freqtrade is a perfect match for this reinforcing mechanism since this information is readily available in live deployments.

What does it mean ? Does it mean that Backtesting is not supported with LR ?

### Relevant code exceptions or logs

  ```
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:10,489 - freqtrade.data.dataprovider - INFO - Increasing startup_candle_count for freqai to 43220
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:13,888 - freqtrade.data.dataprovider - INFO - Increasing startup_candle_count for freqai to 43220
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:15,750 - freqtrade.data.dataprovider - INFO - Increasing startup_candle_count for freqai to 43220
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:17,430 - freqtrade.data.dataprovider - INFO - Increasing startup_candle_count for freqai to 43220
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:17,462 - freqtrade.data.history.idatahandler - WARNING - ETH/USDT:USDT, futures, 1h, data starts at 2020-01-01 00:00:00
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:18,992 - freqtrade.data.dataprovider - INFO - Increasing startup_candle_count for freqai to 43220
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:19,004 - freqtrade.data.history.idatahandler - WARNING - ETH/USDT:USDT, futures, 4h, data starts at 2020-01-01 00:00:00
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:21,625 - freqtrade.freqai.freqai_interface - INFO - Could not find model at /user_data/models/ai_rl_01/sub-train-MATIC_1672531200/cb_matic_1672531200
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:21,625 - freqtrade.freqai.RL.BaseReinforcementLearningModel - INFO - --------------------Starting training MATIC/USDT:USDT --------------------
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:23,072 - freqtrade.freqai.data_kitchen - INFO - MATIC/USDT:USDT: dropped 2 training points due to NaNs in populated dataset 43200.
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:41,206 - freqtrade.freqai.RL.BaseReinforcementLearningModel - INFO - Training model on 1131 features and 28942 data points
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:41,206 - freqtrade.freqai.RL.BaseEnvironment - WARNING - add_state_info is not available in backtesting. Deactivating.
freqtrade-train_ai_rl_01  | 2023-03-29 15:29:41,206 - freqtrade.freqai.RL.BaseEnvironment - WARNING - add_state_info is not available in backtesting. Deactivating.
  ```

Cheers,
Escanor

**Comment by robcaulk** [comment]
Hello,

Backtesting is available in RL, but you just cannot `add_state_info`. This limitation is due to Freqtrade vectorized optimizations in backtesting.

> I was expecting that after the warning, freqtrade continues

You expected correctly, and therefore there could be a bug. I havent noticed anyone else in the discord mentioning issues with backtesting, so it seems odd. But, I will investigate as soon as I have a moment, then I will report back to this thread.

Cheers,

Robert

**Comment by robcaulk** [comment]
I cannot reproduce your bug. I am able to run backtest with RL. Can you please try with a native install instead of using the docker image?

**Comment by escanoro** [comment]
Can you share the strategy you are using? I'm affraid my strategy is causing the issue.

Okay, i will try to have a fresh non docker installation and test it.

I want to note that using the freqai-rl stable image throws additional errors, this is why i switched to develop version.

Do you want me to share the errors too?

One error i recall is that `purge_old_models` required a `bool` and not an `integer`.

Cheers,
Escanor

**Comment by xmatthias** [comment]
`purge_old_models` WAS changed from bool to integer (with backwards compatibility - see [changelog of 2023.2](https://github.com/freqtrade/freqtrade/releases/tag/2023.2)) - but it's been a boolean only in 2023.1 and prior - which would mean your "stable" image is outdated by 1 version (maybe attempt a new `docker-compose pull` to ensure you have an actual, uptodate version)?

**Comment by escanoro** [comment]
Same issue !

> freqtrade-train_ai_rl_01  | Failed validating 'type' in schema['properties']['freqai']['properties']['purge_old_models']:
> freqtrade-train_ai_rl_01  |     {'default': True, 'type': 'boolean'}
> freqtrade-train_ai_rl_01  | 
> freqtrade-train_ai_rl_01  | On instance['freqai']['purge_old_models']:
> freqtrade-train_ai_rl_01  |     5
> freqtrade-train_ai_rl_01  | 2023-03-30 12:51:17,978 - freqtrade - ERROR - Fatal exception!
> freqtrade-train_ai_rl_01  | Traceback (most recent call last):
> freqtrade-train_ai_rl_01  |   File "/freqtrade/freqtrade/main.py", line 42, in main
> freqtrade-train_ai_rl_01  |     return_code = args['func'](args)
> freqtrade-train_ai_rl_01  |   File "/freqtrade/freqtrade/commands/optimize_commands.py", line 52, in start_backtesting
> freqtrade-train_ai_rl_01  |     config = setup_optimize_configuration(args, RunMode.BACKTEST)
> freqtrade-train_ai_rl_01  |   File "/freqtrade/freqtrade/commands/optimize_commands.py", line 21, in setup_optimize_configuration
> freqtrade-train_ai_rl_01  |     config = setup_utils_configuration(args, method)
> freqtrade-train_ai_rl_01  |   File "/freqtrade/freqtrade/configuration/config_setup.py", line 25, in setup_utils_configuration
> freqtrade-train_ai_rl_01  |     validate_config_consistency(config, preliminary=True)
> freqtrade-train_ai_rl_01  |   File "/freqtrade/freqtrade/configuration/config_validation.py", line 96, in validate_config_consistency
> freqtrade-train_ai_rl_01  |     validate_config_schema(conf, preliminary=preliminary)
> freqtrade-train_ai_rl_01  |   File "/freqtrade/freqtrade/configuration/config_validation.py", line 66, in validate_config_schema
> freqtrade-train_ai_rl_01  |     raise ValidationError(
> freqtrade-train_ai_rl_01  | jsonschema.exceptions.ValidationError: 5 is not of type 'boolean'



i don't know if this related , but everytime i pull the freqtradeorg/freqtrade:stable_freqairl, i always get a different Digest:
![image](https://user-images.githubusercontent.com/128816061/228834272-e8db1f0e-b9d4-4f2d-b3a2-0b125f5618de.png)


The current digest is dockerhub is: sha256:29fe8b5fcad65f4552f548eb4dde179aa7582a0c6607cc6200e040b9f8923e76

![image](https://user-images.githubusercontent.com/128816061/228842496-b5db7690-0def-453b-8f00-2b5991a84b9d.png)


I have made this a 3 differnet machines.

![image](https://user-images.githubusercontent.com/128816061/228842256-a8995c82-baa6-48b1-b602-b334edd3f6c7.png)


Regards,
Escanor

**Comment by xmatthias** [comment]
The digest is not necessarily relevant, and it's a somewhat (un)known (bug or feature) that the digest shown on the docker platform is different from the one you get when pulling.

What's the output of `docker compose run --rm freqtrade -V` ?

**Comment by escanoro** [comment]
On both machines (My machine and a new freshly Ubuntu), i got the same version

```
sam@test01:~$ docker compose run --rm freqtrade-train_ai_rl_01 -V
freqtrade 2023.1
```

My compose file
---
```
---
version: '3'
services:
  freqtrade-train_ai_rl_01:
    image: freqtradeorg/freqtrade:stable_freqairl
    restart: "no"
    container_name: freqtrade-train_ai_rl_01
    command: >
      -V
```

**Comment by xmatthias** [comment]
Interesting, you're right, the stable freqAIrl image seems to be containing 2023.1 instead of 2023.2.

i'll look into this when i find the time.

**Comment by xmatthias** [comment]
Fixed the images, stable is now 2023.3 (the new release).

**Comment by escanoro** [comment]
> I cannot reproduce your bug. I am able to run backtest with RL. Can you please try with a native install instead of using the docker image?

The issue was with my stratgey. 

Thanks,
Regards,
Escanor
